### DISCUSSION TITLE
Set up AOS within app directory.

### DISCUSSION THREAD

**Discussion by vixero-dev** [vercel/next.js#49021] [Help]
### Summary

I started learning next.js and wanted to experiment in the app directory right away.

I'm having trouble setting up the [aos](https://github.com/michalsnik/aos) library.

### Additional information

This is the Aos component.

```typescript

'use client';

import React, { useEffect } from 'react';
import AOS from 'aos';

export default function Aos() {
  useEffect(() => {
    AOS.init({
      duration: 1200,
    });
  }, []);

  return <></>;
}
```

And the `layout.tsx`

```typescript
import { ReactNode } from 'react';

import Aos from '@shared/animated-on-scroll';

import '../styles.scss';

export default function RootLayout({ children }: { children: ReactNode }) {
  return (
    <>
      <Aos />
      <html lang="en">
        <body>
          {children}
        </body>
      </html>
    </>
  );
}
```

The error I'm getting:
![image](https://user-images.githubusercontent.com/12700162/235376206-45f98a62-f348-4788-b46f-7ea9bc18b032.png)

There is also a delay of about 1s when data attributes are attached to the `body`-tag, which make the initial animation out of sync. The subsequent animations work as expected.

![image](https://user-images.githubusercontent.com/12700162/235376253-3939aba2-fe5d-4f0e-bf7e-22fd2bbfb5cb.png)

My assumption is that the server side component is being pushed to the client and then the client component gets rendered.

Strangely enough, if I were to delete the `loading.tsx` file, the error from the console disappears, only the delay persists.

After some digging I found [how to load external libraries](https://beta.nextjs.org/docs/optimizing/lazy-loading#example-loading-external-libraries), but I get a runtime error when I try it.

```typescript
'use client';

import React, { useEffect } from 'react';
const AOS = (await import('aos')).default;

export default function Aos() {
  useEffect(() => {
    AOS.init({
      duration: 1200,
    });
  }, []);

  return <></>;
}
```

![image](https://user-images.githubusercontent.com/12700162/235376796-f74b47df-cb6f-400d-a9e1-a290bb8db5d1.png)

### Reproduction

https://stackblitz.com/edit/vercel-next-js-f7x6kq

**Comment by icyJoseph** [comment]
Hi,

You had it right the first time, but you got played by the `loading.tsx` file you've got. If you remove it the, class name divergence should stop. This has to do with how the DOM is organized as things are loaded with suspense boundaries in the mix.

- Try renaming `loading.tsx` to `_loading.tsx` to prevent it from being used.

So the AOS component should've been part of the page, rather than the layout file.

Your interpretation of dynamically loading libraries was a bit off:

```js
'use client';

import { useEffect } from 'react';

export default function Aos() {
  useEffect(() => {
    import('aos').then((AOS) =>
      AOS.init({
        duration: 1200,
      })
    );
  }, []);

  return <></>;
}

```

That's how it should've been. And IIRC, that'll also make it work with the `loading.tsx` file included, but I believe there's a race condition there. 

**If AOS loads faster than the `loading.tsx` is swapped by the home page component, you'd see the error again.**

I think in this case, `templates` are the way to go, but I am not sure. The root solution is to initialise AOS inside Home:

- Templates: https://beta.nextjs.org/docs/routing/pages-and-layouts#templates

```js
import Aos from './aos';

/** Add your relevant code here for the issue to reproduce */
export default function Home() {
  return (
    <>
      <Aos />
      <div
        className="container-fluid main-container container-home p-0"
        data-aos="fade-up"
        data-aos-duration="1200"
      >
        Test
      </div>
    </>
  );
}
```

However, doing this for every page feels repetitive and error prone, and I believe that's where templates kick in... 

> You also don't need to be importing React all of the time, Next.js has always handled that for you, and since React 17 most frameworks do too. I mean that `import React, { useState } from 'react';`, should just be `import { useState } from 'react';`

**Comment by vixero-dev** [comment]
Hi @icyJoseph, thank you for taking your time to answer my question.

You are correct, there is indeed a race condition happening only in dev mode. I have tried templates, but wasn't successful. I will wait until there is hopefully more documentation around templates.

**Reply by gursimarsethi** [reply]
Have you been able to figure it out?

**Reply by vixero-dev** [reply]
Hi, I haven't touched it since. I think I just removed the loading.tsx all together.
